lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1815,6 +1811,12 @@ emitAssociatedFunctionReference(ADContext &context, SILBuilder &builder,
     taskCallback(task);
     auto *assocFn = task->getAssociatedFunction(kind);
     auto *ref = builder.createFunctionRef(loc, assocFn);
+    // FIXME(201): Handle direct differentiation of reabstraction thunks.
+    // Tentative solution: clone a new reabstraction thunk where function
+    // argument has a `@differentiable` function type.
+    if (originalFn->isThunk() == IsReabstractionThunk) {
+      // Handle here!
```

Comment:
A diagnostic is already emitted when differentiating the reabstraction thunk because the thunk argument function is opaque:
```
[AD] Activity info for $sS2fIegnr_S2fIegyd_TR at (source=0 parameters=(0))
[ACTIVE] %0 = argument of bb0 : $Float                     // user: %3
[NONE] %1 = argument of bb0 : $@callee_guaranteed (@in_guaranteed Float) -> @out Float // user: %5
[ACTIVE]   %2 = alloc_stack $Float                         // users: %3, %8, %5
[ACTIVE]   %4 = alloc_stack $Float                         // users: %6, %7, %5
[VARIED]   %5 = apply %1(%4, %2) : $@callee_guaranteed (@in_guaranteed Float) -> @out Float
[ACTIVE]   %6 = load %4 : $*Float                          // user: %9
[AD] Cloning original @$sS2fIegnr_S2fIegyd_TR to primal @AD__$sS2fIegnr_S2fIegyd_TR__primal_src_0_wrt_0
[AD] Primal-transforming:
  %5 = apply %1(%4, %2) : $@callee_guaranteed (@in_guaranteed Float) -> @out Float

[AD] Active indices: params={0}, results={0}
<unknown>:0: note: opaque non-'@differentiable' function is not differentiable
[AD] Diagnosing non-differentiability for inst
	  %5 = apply %1(%4, %2) : $@callee_guaranteed (@in_guaranteed Float) -> @out Float

while performing differentiation task
	(differentiation_task original=@$sS2fIegnr_S2fIegyd_TR attribute=source 0 wrt 0 jvp @AD__$sS2fIegnr_S2fIegyd_TR__jvp_src_0_wrt_0 vjp @AD__$sS2fIegnr_S2fIegyd_TR__vjp_src_0_wrt_0 invoker=(differentiation_invoker differential_operator=((autodiff_function implicit type='@differentiable (Float) -> Float' location=ind2.swift:7:34 range=[ind2.swift:7:34 - line:7:34]
  (function_conversion_expr implicit type='(Float) -> Float' location=ind2.swift:7:34 range=[ind2.swift:7:34 - line:7:34]
    (declref_expr type='(Float) -> Float' location=ind2.swift:7:34 range=[ind2.swift:7:34 - line:7:34] decl=ind2.(file).direct@ind2.swift:3:6 [with (substitution_map generic_signature=<T where T : Differentiable, T.CotangentVector : AdditiveArithmetic, T.TangentVector : AdditiveArithmetic> (substitution T -> Float))] function_ref=unapplied))))))
ind2.swift:7:34: error: function is not differentiable
print(pullback(at: Float(1), in: direct)(1))
                                 ^~~~~~
<unknown>:0: note: expression is not differentiable
```

I mainly marked this spot so that "direct differentiation of indirect functions" can be handled in a follow-up (addressing TF-201). Let me know if you feel like we should explicitly emit an error here for now.